refactor(devtools): add matcher and runGuardsAndResolvers support in …

SkyZeroZx opened the pull request:
…router viewer

Enhance the Angular DevTools router viewer to display routes that use custom `matcher` functions and reflect the `runGuardsAndResolvers` configuration

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/contributing-docs/commit-message-guidelines.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.dev application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?
<img width="359" height="296" alt="image" src="https://github.com/user-attachments/assets/fbfaa383-f32e-4e1a-97c3-a80073f06db3" />
<img width="356" height="325" alt="image" src="https://github.com/user-attachments/assets/f7113867-25e0-4937-8284-c8a071e4d33a" />


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

Question : 
When a `matcher` is used, the `path` property is not allowed. However, in the router viewer, it currently displays as `/undefined`.
We should consider replacing `/undefined` with a more descriptive label.
<img width="272" height="287" alt="image" src="https://github.com/user-attachments/assets/a6f20877-8364-4912-af95-a3496b8bec32" />

SkyZeroZx commented:
> #64938 isn't even merged, that changes is too premature ATM.

As I understand it, `runGuardsAndResolvers` (which currently supports a function with `from` and `to` [RunGuardsAndResolvers](https://github.com/angular/angular/blob/main/packages/router/src/models.ts#L339) ) has existed for a while, just like `matcher`.  
The DevTools simply expose its metadata to make it easier to inspect in the router viewer.  
The PR (from what I briefly reviewed) adds DI support for `runGuardsAndResolvers` functions, which remains backward-compatible with the existing behavior.

JeanMeche commented:
I shouldn't be reviewing PRs so late, you're right.

hawkgs commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree.component.html:
```diff
@@ -176,6 +188,19 @@ <h2 class="router-title">Routes Details</h2>
+                  dataKey="runGuardsAndResolvers"
+                  [data]="data"
+                  [actionBtnType]="hasStaticOptionRunGuardsAndResolvers() ? 'view-source' : 'none'"
+                  [actionBtnTooltip]="hasStaticOptionRunGuardsAndResolvers() ? 'View source' : 'none'"
```
```suggestion
                  actionBtnTooltip="View source"
```

No need to have the check. The tooltip is visible only if there is a button anyway.

SkyZeroZx replied:
I think it's necessary because if we set `[actionBtnType]="'view-source'"` directly, it will always show like this when that value is a static string.
For now, the only way to avoid that behavior is by using this flag.

<img width="301" height="337" alt="image" src="https://github.com/user-attachments/assets/c81bf5e9-9954-4560-ae66-545bf1743f9c" />

hawkgs commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree.component.ts:
```diff
@@ -69,6 +76,10 @@ export class RouterTreeComponent {
 
+  protected hasStaticOptionRunGuardsAndResolvers = computed(
+    () => !RUN_GUARDS_AND_RESOLVERS_OPTIONS.includes(this.routeData()?.runGuardsAndResolvers || ''),
+  );
```
The name is actually opposite of the check. Either we need to change it or drop the `!`.

hawkgs commented:
> Question :
> When a matcher is used, the path property is not allowed. However, in the router viewer, it currently displays as /undefined.
> We should consider replacing /undefined with a more descriptive label.

Yeah, that's something that we should handle.

SkyZeroZx commented:
> > Question :
> > When a matcher is used, the path property is not allowed. However, in the router viewer, it currently displays as /undefined.
> > We should consider replacing /undefined with a more descriptive label.
> 
> Yeah, that's something that we should handle.


I think something like this is useful to indicate that when using a matcher or other router details, the custom name should be used if it exists.
Avoid relying on the function’s name, because inline functions may not have one.

<img width="277" height="272" alt="image" src="https://github.com/user-attachments/assets/5e4faf76-c16f-4e48-9a2c-83e9e09e4027" />

SkyZeroZx committed:
refactor(devtools): add matcher and runGuardsAndResolvers support in router viewer

Enhance the Angular DevTools router viewer to display routes that use custom `matcher` functions and reflect the `runGuardsAndResolvers` configuration
